[Screen Time Refactoring] Currently blocked web content briefly visible when navigating away

jesxilin opened the pull request:
#### 2ea18446ee3b5b205d2d5abc6490fd314c4cc1f8
<pre>
[Screen Time Refactoring] Currently blocked web content briefly visible when navigating away
<a href="https://bugs.webkit.org/show_bug.cgi?id=284643">https://bugs.webkit.org/show_bug.cgi?id=284643</a>
<a href="https://rdar.apple.com/141447068">rdar://141447068</a>

Reviewed by Wenson Hsieh.

Currently, switching from blocked URL to unblocked URL will show web content
of blocked URL. This is due to the URL being changed and processed, thus hiding the blocking
view before the new web content has been displayed.
To address this issue, we will now update the ScreenTimeWebpageController&apos;s
URL only after a webpage has updated display.

* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didChangeScreenTimeWebpageControllerURL):
(WebKit::PageClientImplCocoa::updateScreenTimeWebpageControllerURL):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didCommitLayerTree):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::didChangeScreenTimeWebpageControllerURL):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didCommitLoadForMainFrame):
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::didCommitLoadForMainFrame):

Canonical link: <a href="https://commits.webkit.org/287961@main">https://commits.webkit.org/287961@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/622a490b495a16cc9f425002827bdc6cd44a6128

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81294 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/819 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35237 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85823 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32280 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/837 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8634 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63461 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21229 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84363 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/567 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73985 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43758 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/465 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30738 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71960 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28727 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87258 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8524 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6057 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71772 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8705 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69815 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71006 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17715 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15061 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13973 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8486 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14009 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8322 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11843 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10130 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
```diff
@@ -345,6 +345,10 @@
     }
+
+    if (pageProxy)
+        pageProxy->makeViewBlankIfUnpaintedSinceLastLoadCommit();
```
Isn't this going to make the view blank every time you navigate to a new page, regardless of whether the blocking view was shown?

jesxilin replied:
Yeah, I was just about to talk to you about it lol. I think maybe smth like if URLIsBlocked --> make view blank before navigating

jesxilin replied:
I also was gonna ask about the pageProxy check (since I check it before as well so it's redundant)

whsieh replied:
> I also was gonna ask about the pageProxy check

You can probably return early (uninstalling the ST controller if needed), if the page is null.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
```diff
@@ -321,6 +321,34 @@
+    }
+
+    if (pageProxy)
+        pageProxy->makeViewBlankIfUnpaintedSinceLastLoadCommit();
```
Also, we shouldn't use `makeViewBlankIfUnpaintedSinceLastLoadCommit` without making refactoring to that method. Currently, it is specifically for JS prompts, as it has logging about that and is also conditional based on a linked-on-or-after check.

hortont424 commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
```diff
@@ -330,6 +330,10 @@
 {
     [m_webView _installScreenTimeWebpageController];
 }
+void PageClientImplCocoa::didChangeScreenTimeWebpageControllerURL()
```
newline above here

hortont424 commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -172,8 +172,12 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+    if (RefPtr pageClient = this->pageClient()) {
         pageClient->didCommitLayerTree(layerTreeTransaction);
+#if ENABLE(SCREEN_TIME)
+        pageClient->didChangeScreenTimeWebpageControllerURL();
```
This doesn't seem like the right place. You're doing this every single commit, which happens a lot. Don't you want it up next to `m_hasUpdatedRenderingAfterDidCommitLoad = true`, so you only do it for the first remote layer tree commit after didCommitLoad?

jesxilin replied:
I believe you are right. Will change now!

hortont424 reviewed (commented):
The title should describe the problem, not say what you're fixing.

jesxilin committed:
[Screen Time Refactoring] Currently blocked web content briefly visible when navigating away
https://bugs.webkit.org/show_bug.cgi?id=284643
rdar://141447068

Reviewed by Wenson Hsieh.

Currently, switching from blocked URL to unblocked URL will show web content
of blocked URL. This is due to the URL being changed and processed, thus hiding the blocking
view before the new web content has been displayed.
To address this issue, we will now update the ScreenTimeWebpageController's
URL only after a webpage has updated display.

* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didChangeScreenTimeWebpageControllerURL):
(WebKit::PageClientImplCocoa::updateScreenTimeWebpageControllerURL):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didCommitLayerTree):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::didChangeScreenTimeWebpageControllerURL):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didCommitLoadForMainFrame):
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::didCommitLoadForMainFrame):

Canonical link: https://commits.webkit.org/287961@main
